docs/developers/functional-tests.md

```diff
@@ -0,0 +1,114 @@
+## Functional tests
+
+Main language: Groovy. Project functional tests use [Spock](https://spockframework.org/) as a main testing framework.
+Also used [Docker](https://www.docker.com/) for running PBS and other services.
+And [MockServer](https://www.mock-server.com/) for mocking external services.
+
+## Getting Started
+
+- Install [Docker](https://docs.docker.com/engine/install/)
+- Set launchContainers system property to 'true'
+
+### Prepare PBS image
+
+`mvn clean -B package -DskipUnitTests=true docker:build` - to build image
+
+> Don't forget to rebuild image for manual test start (after change in dev code)
+
+### Run integration tests
+
+- You can use `mvn verify` - to include all previous steps (including java test) because groovy runs
+  in `failsafe:integration-test` phase
+- Or use more granular `mvn -B verify -DskipUnitTests=true` for functional tests only
+
+## Developing
+
+Functional tests need to have name template **.\*Spec.groovy**
+
+- `/functional/*Spec` - contain all test suites.
+- `/functional/util` - different useful methods and wrappers.
+- `/functional/repository` - interaction with DB. Notice, that some related objects need to use ID generated by DB.
+  Primary Key will be inserted after saving instance into DB.
+- `/functional/service/PrebidServerService` - responsible for all PBS http calls.
+- `/functional/testcontainers/Dependencies` - stores dependencies and manages mySql and NetworkServiceContainer containers.
+- `/functional/testcontainers/ErrorListener` - logs request and response in case of falling test.
+- `/functional/testcontainers/PbsConfig` - collects PBS properties.
+- `/functional/testcontainers/PbsServiceFactory` - manage PBS containers according to container limit.
+- `/functional/testcontainers/PBSTestExtension` - allows to hook into a spec’s lifecycle to add ErrorListener using annotation `PBSTest`.
+- `/functional/testcontainers/TestcontainersExtension` - allow to hook into a spec’s lifecycle to start and stop support service containers using global extension.
+- `/functional/testcontainers/container` - responsible for creating and configuring containers.
+- `/functional/testcontainers/scaffolding/NetworkScaffolding` -  makes HTTP requests to a MockServer.
+
+
+**Properties:**
+
+`launchContainers` - responsible for starting the MockServer and the MySQLContainer container. Default value is false.
```

Comment:
Please leave a note that default value is false to not launch containers for unit tests.